### PR TITLE
[NF] DAM implementation for tissue classification using DMRI signal properties.

### PR DESCRIPTION
**Pull request by Atharva-Shah-2298**
This PR is for addition of tissue classifier method based on Directional Average maps based on the paper

Cheng, H., Newman, S., Afzali, M., Fadnavis, S. S., & Garyfallidis, E. (2020). Segmentation of the brain using direction-averaged signal of DWI images. Magnetic Resonance Imaging, 69, 1–7. https://doi.org/10.1016/j.mri.2020.02.010 .

Usage Example:

```
from dipy.io.gradients import read_bvals_bvecs
from dipy.io.image import load_nifti
from dipy.core.gradients import gradient_table
from dipy.segment.mask import median_otsu
import matplotlib.pyplot as plt
from dipy.segment.tissue import dam_classifier

data, affine = load_nifti("100307/T1w/Diffusion/data.nii.gz")

bvals, bvecs = read_bvals_bvecs("100307/T1w/Diffusion/bvals", "100307/T1w/Diffusion/bvecs")
gtab = gradient_table(bvals, bvecs)

maskdata, mask = median_otsu(data, vol_idx=[0,1], median_radius=2, numpass=5,
                             autocrop=False, dilate=1)

wm_mask, gm_mask = dam_classifier(data, bvals, wm_threshold=0.5, b0_thresh=50, low_signal_threshold=50)

plt.figure(figsize=(10, 5))
plt.subplot(1, 2, 1)
plt.imshow(wm_mask[:, :, data.shape[2]//2], cmap='gray')
plt.title('White Matter Mask')
plt.axis('off')

plt.subplot(1, 2, 2)
plt.imshow(gm_mask[:, :, data.shape[2]//2], cmap='gray')
plt.title('Grey Matter Mask')
plt.axis('off')

plt.show()
```
The results for a HCP data sample

image

### CODE REVIEW

File: dipy/segment/tissue.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
 from dipy.segment.mrf import ConstantObservationModel, IteratedConditionalModes
 from dipy.sims.voxel import add_noise
 from dipy.testing.decorators import warning_for_keywords
-
+from sklearn.linear_model import LinearRegression
```
Comment (skoudoro) [inline dipy/segment/tissue.py:6]: scikit-learn is an optional package. Please, look at the codebase to get some inspiration of `optional_package` use.

File: dipy/segment/tissue.py [hunk 2]
```diff
@@ -125,3 +125,98 @@ def classify(self, image, nclasses, beta, *, tolerance=1e-05, max_iter=100):
         PVE = PVE[..., 1:]
 
         return initial_segmentation, final_segmentation, PVE
+
+
+def compute_directional_average(data, bvals, b0_thresh=10):
```
Comment (skoudoro) [inline dipy/segment/tissue.py:130]: - `b0_threshold=50` instead - to be consistent with the codebase
- Also, we use keyword only arguments so the definition should be: 
   `def compute_directional_average(data, bvals, *, b0_threshold=50):`

File: dipy/segment/tissue.py [hunk 3]
```diff
@@ -125,3 +125,98 @@ def classify(self, image, nclasses, beta, *, tolerance=1e-05, max_iter=100):
...
+    """
+    Compute the mean signal for each unique b-value shell and fit a linear model.
+
+    Parameters
+    ----------
+        data : ndarray
+            The diffusion MRI data.
+        bvals : ndarray
+            The b-values corresponding to the diffusion data.
+
```
Comment (skoudoro) [inline dipy/segment/tissue.py:140]: missing `b0_threshold`

File: dipy/segment/tissue.py [hunk 4]
```diff
@@ -125,3 +125,98 @@ def classify(self, image, nclasses, beta, *, tolerance=1e-05, max_iter=100):
...
+    """
+    unique_bvals = np.unique(bvals)
+    # Mean signal for b=0 (non-diffusion weighted)
+    s0 = data[bvals < b0_thresh].mean()
+
+    if len(unique_bvals) <= 2:
+        raise ValueError("Insufficient unique b-values for fitting.")
+
+    # If the mean signal for b=0 is too low, return 0, 0
+    if data[bvals < b0_thresh].mean() < 50:
```
Comment (skoudoro) [inline dipy/segment/tissue.py:156]: why 50 ? it should be documented in the docstring. Or maybe it can be an input parameter for the user.

File: dipy/segment/tissue.py [hunk 5]
```diff
@@ -125,3 +125,98 @@ def classify(self, image, nclasses, beta, *, tolerance=1e-05, max_iter=100):
...
+        return 0, 0
+
+    S_bvals = [data[bvals == bval].mean() / (s0 + 0.01)
+               for bval in unique_bvals[1:]]
+    S_bvals = np.array(S_bvals)
```
Comment (skoudoro) [inline dipy/segment/tissue.py:161]: a loop can be avoided here since we know the shape of the final array. Can you improve this?

File: dipy/segment/tissue.py [hunk 6]
```diff
@@ -125,3 +125,98 @@ def classify(self, image, nclasses, beta, *, tolerance=1e-05, max_iter=100):
...
+
+    # Avoid log(0)
+    if 0 in S_bvals:
+        S_bvals = S_bvals + 0.001
+
+    S_log = np.log(S_bvals)
```
Comment (skoudoro) [inline dipy/segment/tissue.py:167]: `S_log = np.log(S_bvals, out=np.zeros_like(S_bvals, dtype=np.float64), where=(S_bvals != 0))`. instead of adding 0.001

File: dipy/segment/tissue.py [hunk 7]
```diff
@@ -125,3 +125,98 @@ def classify(self, image, nclasses, beta, *, tolerance=1e-05, max_iter=100):
...
+    # Fit linear model
+    model = LinearRegression()
+    model.fit(X, y)
+    P = model.coef_[0]
+    V = model.intercept_
+
+    return P, V
+
+
+def compute_P_map(data, bvals, wm_threshold=0.5, b0_thresh=10):
```
Comment (skoudoro) [inline dipy/segment/tissue.py:184]: same comments as above concerning ` b0_thresh` and keyword only arguments.

Also, p should be lowercase in the function name

File: dipy/segment/tissue.py [hunk 8]
```diff
@@ -125,3 +125,98 @@ def classify(self, image, nclasses, beta, *, tolerance=1e-05, max_iter=100):
...
+
+    Parameters
+    ----------
+        data : ndarray
+            The diffusion MRI data.
+        bvals : ndarray
+            The b-values corresponding to the diffusion data.
+        wm_threshold : float, optional
+            The threshold below which a voxel is considered white matter.
+            Default is 0.5.
```
Comment (skoudoro) [inline dipy/segment/tissue.py:196]: default to remove

File: dipy/segment/tissue.py [hunk 9]
```diff
@@ -125,3 +125,98 @@ def classify(self, image, nclasses, beta, *, tolerance=1e-05, max_iter=100):
...
+        b0_thresh : float, optional
+            The threshold below for a b=0 image.
+            Default is 10.
```
Comment (skoudoro) [inline dipy/segment/tissue.py:199]: default to remove

File: dipy/segment/tissue.py [hunk 10]
```diff
@@ -125,3 +125,98 @@ def classify(self, image, nclasses, beta, *, tolerance=1e-05, max_iter=100):
...
         return initial_segmentation, final_segmentation, PVE
+
+
+def compute_directional_average(data, bvals, b0_thresh=10):
+    """
+    Compute the mean signal for each unique b-value shell and fit a linear model.
+
+    Parameters
+    ----------
+        data : ndarray
```
Comment (skoudoro) [inline dipy/segment/tissue.py:136]: indentation is wrong. Can you align it to Parameters. (see other docstring in the codebase)

File: dipy/segment/tissue.py [hunk 11]
```diff
@@ -125,3 +125,98 @@ def classify(self, image, nclasses, beta, *, tolerance=1e-05, max_iter=100):
...
+            The diffusion MRI data.
+        bvals : ndarray
+            The b-values corresponding to the diffusion data.
+
+    Returns
+    -------
+        P : float
+            The slope of the linear model.
+        V : float
+            The intercept of the linear model.
```
Comment (skoudoro) [inline dipy/segment/tissue.py:146]: indentation is wrong. Can you align it to Returns. (see other docstring in the codebase)

File: dipy/segment/tissue.py [hunk 12]
```diff
@@ -125,3 +125,98 @@ def classify(self, image, nclasses, beta, *, tolerance=1e-05, max_iter=100):
...
+    Compute the P (slope) map for the entire dataset.
+
+    Parameters
+    ----------
+        data : ndarray
+            The diffusion MRI data.
+        bvals : ndarray
+            The b-values corresponding to the diffusion data.
+        wm_threshold : float, optional
+            The threshold below which a voxel is considered white matter.
```
Comment (skoudoro) [inline dipy/segment/tissue.py:195]: indentation is wrong. Can you align it to Parameters. (see other docstring in the codebase)

File: dipy/segment/tissue.py [hunk 13]
```diff
@@ -125,3 +125,98 @@ def classify(self, image, nclasses, beta, *, tolerance=1e-05, max_iter=100):
...
+        b0_thresh : float, optional
+            The threshold below for a b=0 image.
+            Default is 10.
+
+    Returns
+    -------
+        wm_mask : ndarray
+            A binary mask for white matter.
+        gm_mask : ndarray
+            A binary mask for grey matter.
```
Comment (skoudoro) [inline dipy/segment/tissue.py:206]: indentation is wrong. Can you align it to Returns. (see other docstring in the codebase)

File: dipy/segment/tissue.py [hunk 14]
```diff
@@ -125,3 +125,98 @@ def classify(self, image, nclasses, beta, *, tolerance=1e-05, max_iter=100):
...
+    P_map = np.zeros(data.shape[:-1])
+    for i in range(data.shape[0]):
+        for j in range(data.shape[1]):
+            for k in range(data.shape[2]):
+                voxel_data = data[i, j, k, :]
+                if np.any(bvals != 0):
+                    P, _ = compute_directional_average(voxel_data, bvals, b0_thresh)
+                    P_map[i, j, k] = P
+
+    wm_mask = (P_map <= wm_threshold) & (P_map > 0.01)
```
Comment (skoudoro) [inline dipy/segment/tissue.py:217]: 0.01 ?  magic number ? Can you explain or add a note in the docstring

**Review by skoudoro — CHANGES_REQUESTED** [review]
Hi @Atharva-Shah-2298,

Thank you for this addition. See below my initial review. More to come after addressing this review.

**Comment by skoudoro** [comment]
Also, it would be great to:
- add some tests
- update the tutorial using TissueClassifier 

thank you

File: dipy/segment/tissue.py (hunk 6 above)
Comment (Atharva-Shah-2298) [reply dipy/segment/tissue.py:167]: I see what you mean but the log of a number near zero would be a big negative number and not zero. So adding 1e-3 would be sensible.

File: dipy/segment/tissue.py (hunk 4 above)
Comment (Atharva-Shah-2298) [reply dipy/segment/tissue.py]: Added extra low signal threshold for users but kept it as 50 default as It is a tricky parameter to identify.

File: dipy/segment/tissue.py (hunk 1 above)
Comment (Atharva-Shah-2298) [reply dipy/segment/tissue.py]: Added optional_package import

File: dipy/segment/tissue.py [hunk 15]
```diff
@@ -125,3 +126,115 @@ def classify(self, image, nclasses, beta, *, tolerance=1e-05, max_iter=100):
...
+
+    Parameters
+    ----------
+    data : ndarray
+        The diffusion MRI data.
+    bvals : ndarray
+        The b-values corresponding to the diffusion data.
+    b0_thresh : float, optional
+        The threshold below for a b=0 image.
+        Default is 50.
```
Comment (jhlegarreta) [inline dipy/segment/tissue.py:144]: Please, avoid saying here what the default value is: it is clear from the method signature. If the value in the signature is changed for whatever reasons, the risk of not updating the parameter docstring value is high, and so the documentation would go out of sync, be misleading, and require more maintenance. Remove all values except for those cases where the default value requires some coding in the method making the parameter take a value that is not the same as the one in the signature.

**Review by jhlegarreta — CHANGES_REQUESTED** [review]

File: dipy/segment/tissue.py (hunk 15 above)
Comment (Atharva-Shah-2298) [reply dipy/segment/tissue.py:144]: Yeah. Makes sense. Thank you for the comment. Will remove it asap

**Comment by Garyfallidis** [comment]
@ShreyasFadnavis can you look at this PR?

**Comment by Atharva-Shah-2298** [comment]
Also, the method requires multiple unique bvalues to fit and acquisitions with lower number of unique bvalues maybe prone to underfitting.

**Comment by skoudoro** [comment]
@ShreyasFadnavis,  Can you have a look at this PR please ? Do you have an idea when you will have the time to do it ?

File: dipy/segment/tissue.py [hunk 16]
```diff
@@ -125,3 +129,136 @@ def classify(self, image, nclasses, beta, *, tolerance=1e-05, max_iter=100):
...
+):
+    """Compute the P (slope) map for the entire dataset.
+
+    Parameters
+    ----------
+    data : ndarray
+        The diffusion MRI data.
+    bvals : ndarray
+        The b-values corresponding to the diffusion data.
+    wm_threshold : float, optional
```
Comment (skoudoro) [inline dipy/segment/tissue.py:217]: Can you remove the `,optional`

File: dipy/segment/tissue.py [hunk 17]
```diff
@@ -125,3 +129,136 @@ def classify(self, image, nclasses, beta, *, tolerance=1e-05, max_iter=100):
...
+        The diffusion MRI data.
+    bvals : ndarray
+        The b-values corresponding to the diffusion data.
+    s0_map : ndarray, optional
+        Precomputed mean signal map for b=0 images.
+    masks : ndarray, optional
+        Precomputed masks for each unique b-value shell.
+    b0_threshold : float, optional
+        The intensity threshold for a b=0 image.
+    b0_mask : ndarray, optional
```
Comment (skoudoro) [inline dipy/segment/tissue.py:159]: Can you change the docstring order to respect what is above

**Review by skoudoro — COMMENTED** [review]
Hi @Atharva-Shah-2298,

Overall, it looks good.

Can you add the citation: 
- first, you need to add it herehttps://github.com/dipy/dipy/blob/master/doc/references.bib
- please, respect the alphabetical order you add the citation
- Then, you need to add `footcite` and  `..footbibliography` in the docstring. see an example here: https://github.com/dipy/dipy/blob/2a747d9590f967d5e9cde849dde4cb539d91a6d5/dipy/denoise/gibbs.py#L147-L166

thank you

**Comment by ShreyasFadnavis** [comment]
Hey @Atharva-Shah-2298 ! Thanks for doing this - I dint get a chance to test it too much, but the code overall looks good!

**Comment by skoudoro** [comment]
Hi @Atharva-Shah-2298,

Why the methods does not return the CSF masks. The paper indicates that it should do it. thank you for your feedback

**Comment by Atharva-Shah-2298** [comment]
> Hi @Atharva-Shah-2298,
> 
> Why the methods does not return the CSF masks. The paper indicates that it should do it. thank you for your feedback

The authors write that they have done CSF segmentation using beta maps but I quote
"However, the segmentation of CSF is not optimal
using hard thresholding. As CSF is mostly adjacent to GM, a small
threshold of beta can lead to more inclusion of GM in CSF classification."

While implementation, I found results which matched the concerns of the authors. As CSF is very sensitive to thresholding, I have not implemented it.

**Comment by skoudoro** [comment]
thank you for the feedback !

File: doc/references.bib [hunk 18]
```diff
@@ -312,6 +312,31 @@ @article{Chandio2020a
   url       = {https://doi.org/10.1038/s41598-020-74054-4}
 }
 
+@article{Chang2005,
```
[the remark this answers is not in the document]
Comment (Atharva-Shah-2298) [reply doc/references.bib:315]: Modified the list. Now Cheng 2022 is on top and Cheng 2020 is shifted down

**Review by skoudoro — APPROVED** [review]
All Good, Thank you @Atharva-Shah-2298,

I am going ahead and merge this.

it would be great to have a tutorial for this. I am creating an issue as a reminder.
